{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8222ec7a-6e5a-47ec-bf9c-0b7b9ea75653",
   "metadata": {},
   "source": [
    "# Selecting Logan Search results by Sequencing Technology using Python in Jupyter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5e964cf-0c99-4d4e-bd94-dbac316027aa",
   "metadata": {},
   "source": [
    "This notebook demonstrates how you could filter Logan Search Results by technolgy.\n",
    "\n",
    "This assumes you have worked through [Working with with Logan Search results in Jupyter](Working_with_Logan_Search_results_with_Python_in_Jupyter.ipynb). If you haven't yet, please do that first.   \n",
    "Indeed the first code cell below is directly from the first cell [there](Working_with_Logan_Search_results_with_Python_in_Jupyter.ipynb).\n",
    "\n",
    "\n",
    "----------\n",
    "\n",
    "#### Reading in the results for several queries\n",
    "\n",
    "Run the code in the next cell assigning the settings to match the set-up preparation performed and to bring in the results data to the active namespace:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0c8ecae-adfe-4082-b984-3faa9a60d873",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from human_USP7_TSL4_altSJ_31mer_logan_results.tsv:  19997 rows read in\n",
      "from human_USP7_TSL4_SJ_31mer_logan_results.tsv:     19998 rows read in\n",
      "from human_USP7_SJ_14n15_31mer_logan_results.tsv:    19999 rows read in\n",
      "from human_USP7_SJ_34_31mer_logan_results.tsv:       19997 rows read in\n",
      "\n",
      "TOTAL QUERY RESULTS READ IN: 4\n"
     ]
    }
   ],
   "source": [
    "results_dir = \"/home/jovyan/logan_search_results/\"\n",
    "consistent_part_of_file_name_suffix = \"_31mer_\"\n",
    "\n",
    "import os\n",
    "import fnmatch\n",
    "import pandas as pd\n",
    "\n",
    "def extract_tags_from_filename(filename):\n",
    "    '''\n",
    "    extract tag labels from file names\n",
    "\n",
    "    return the tag label\n",
    "    '''\n",
    "    first_part = filename.split(consistent_part_of_file_name_suffix,1)[0]\n",
    "    all_but_first_part_of_remainder_as_list = first_part.split(\"_\")[1:]\n",
    "    return \"_\".join(all_but_first_part_of_remainder_as_list)\n",
    "    \n",
    "matched_filenames = []\n",
    "file_tags_per_file = []\n",
    "dfs = []\n",
    "for file in os.listdir(results_dir):\n",
    "    if fnmatch.fnmatch(file, '*.tsv'):\n",
    "        matched_filenames.append(file)\n",
    "        file_tags_per_file.append(extract_tags_from_filename(file))\n",
    "        dfs.append(pd.read_csv(results_dir + file, sep=\"\\t\"))\n",
    "\n",
    "# provide feedback about what was read in\n",
    "max_filename_length = max(len(filename) for filename in matched_filenames)\n",
    "for filename, df in zip(matched_filenames, dfs): \n",
    "    print(f'from {filename}:{\" \" * (max_filename_length - len(filename))} {len(df):>6} rows read in')\n",
    "\n",
    "#print(file_tags_per_file) # FOR DEBUGGING, uncomment\n",
    "#print(matched_filenames) # FOR DEBUGGING, uncomment\n",
    "gene_name = file_tags_per_file[0].split(\"_\",1)[0]\n",
    "# make tuples of tag labels and corresponding dataframes\n",
    "tags_and_dataframes = list(zip(file_tags_per_file, dfs)) # tags are first in each two-item tuple, the dataframe is second\n",
    "# make dictionary of tuples of tag labels and corresponding dataframes, integers will be the keys with the tuples as values\n",
    "tags_and_dataframes_dict = dict(zip(range(len(file_tags_per_file)), zip(file_tags_per_file, dfs))) # integers ('index') are keys with tags are first in each two-item tuple, the dataframe is second\n",
    "print(f\"\\nTOTAL QUERY RESULTS READ IN: {len(file_tags_per_file)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1cc58ce-f4b2-412a-8f5c-9011c5d84904",
   "metadata": {},
   "source": [
    "--------\n",
    "\n",
    "## Selecting based on Sequencing Technology\n",
    "\n",
    "#### Step #1 add a type column based on Instrument used\n",
    "\n",
    "In step #1, we'll add and additional column to the data based on the `Instruments` column.  \n",
    "To do that first we have to bring in a table that has the technology instruments.  \n",
    "Run the cell below to read the the comma-separated values file (.csv) that contains the details on the instruments that are seen in the results from Logan Search results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b449e5cc-81c1-4976-b535-e3f84f8745ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "instr_dets_df = pd.read_csv(\"supporting_tables/sequencing_instrument_details_for_all_SRA_options_seen_thus_far.csv\")\n",
    "instrument_to_type_dict = dict(zip(instr_dets_df['Instrument'],instr_dets_df['Read Type']))\n",
    "#instr_dets_df.head() #Uncomment this line to display the first five lines of the dataframe made from `.csv` file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e1d5cf9-f2b0-4201-9657-efef839bcea6",
   "metadata": {},
   "source": [
    "Now we are ready to do the assignment step.\n",
    "\n",
    "The following cell should add a column 'reads_type' based on the `instr_dets` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff3292c9-16d5-4ef0-aba5-f45d527a59ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "for df in dfs:\n",
    "    #df['Read Type'] = df['instrument'].apply(instrument_to_type)\n",
    "    df['reads_type'] = df['instrument'].map(instrument_to_type_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e7a2bc9-15c9-4fb8-8921-db0bcfd958e1",
   "metadata": {},
   "source": [
    "If you later edit this notebook to repeat it with your own results, keep in mind tha assignment step may fail. Whike an effort was made to make it fairly extensive, the table may not be complete at this time. If you see an error here it may be that there is no match for an instrument listed in your results among the list supplied here. To fix it though, you should be able to edit the comma-separated values file [there](supporting_tables/sequencing_instrument_details_for_all_SRA_options_seen_thus_far.csv), save it, and run the code again.\n",
    "\n",
    "Let's run the following code to see that a new 'reads_type' column has been added to the first dataframe:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "389eb1f0-e834-487e-8ebb-c78018cb3eda",
   "metadata": {},
   "source": [
    "Now with that column added, we can use it to subset a dataframe in whcih we are interested."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c4a62c1-b0ba-4db8-8326-5f5d08487604",
   "metadata": {},
   "source": [
    "#### Step #2 Select on the technology type\n",
    "\n",
    "We'll filter the human results for 'alt_SJ' based on if the sequencing reads type is '`Long`'. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "722b29f7-8465-4275-880a-d0f923668c76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There's 9096 human datasets in the 'alt_SJ' results.\n",
      "Out of those 9096, there's 8 'long-read' ones.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>kmer_coverage</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>bioproject_title</th>\n",
       "      <th>bioproject_description</th>\n",
       "      <th>sample_acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>experiment</th>\n",
       "      <th>...</th>\n",
       "      <th>avgspotlen</th>\n",
       "      <th>mbases</th>\n",
       "      <th>biosamplemodel_sam</th>\n",
       "      <th>collection_date_sam</th>\n",
       "      <th>location_count</th>\n",
       "      <th>location_avg_coverage</th>\n",
       "      <th>location_max_coverage</th>\n",
       "      <th>location_min_coverage</th>\n",
       "      <th>location_med_coverage</th>\n",
       "      <th>reads_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>SRR14212343</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA721322</td>\n",
       "      <td>SAMN18714204</td>\n",
       "      <td>RNA-Seq analysis (long-reads) of human parenta...</td>\n",
       "      <td>M238P BRAFV600E mutant melanoma cells sensitiv...</td>\n",
       "      <td>SRS8683487</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX10578990</td>\n",
       "      <td>...</td>\n",
       "      <td>698</td>\n",
       "      <td>11980</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1303</th>\n",
       "      <td>SRR22729431</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA912090</td>\n",
       "      <td>SAMN32218795</td>\n",
       "      <td>Exploring collateral RNA cleavage by LbuCas13a...</td>\n",
       "      <td>LbuCas13a exhibits widespread collateral RNA c...</td>\n",
       "      <td>SRS16133703</td>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>LUMC</td>\n",
       "      <td>SRX18690947</td>\n",
       "      <td>...</td>\n",
       "      <td>382</td>\n",
       "      <td>537</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1484</th>\n",
       "      <td>SRR13781371</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA704727</td>\n",
       "      <td>SAMN18053880</td>\n",
       "      <td>Respiratory syncytial virus induces alternativ...</td>\n",
       "      <td>we examine the effect of RSV infection in huma...</td>\n",
       "      <td>SRS8317380</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX10166867</td>\n",
       "      <td>...</td>\n",
       "      <td>1508</td>\n",
       "      <td>33362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3086</th>\n",
       "      <td>ERR5670887</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJEB42052</td>\n",
       "      <td>SAMEA8515329</td>\n",
       "      <td>Targeting the m6A RNA modification pathway blo...</td>\n",
       "      <td>N6-methyladenosine (m6A) is an abundant intern...</td>\n",
       "      <td>ERS6200127</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>NEW YORK UNIVERSITY SCHOOL OF MEDICINE</td>\n",
       "      <td>ERX5386380</td>\n",
       "      <td>...</td>\n",
       "      <td>1669</td>\n",
       "      <td>753</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3730</th>\n",
       "      <td>SRR13781369</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA704727</td>\n",
       "      <td>SAMN18053882</td>\n",
       "      <td>Respiratory syncytial virus induces alternativ...</td>\n",
       "      <td>we examine the effect of RSV infection in huma...</td>\n",
       "      <td>SRS8317378</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX10166865</td>\n",
       "      <td>...</td>\n",
       "      <td>1481</td>\n",
       "      <td>27528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7840</th>\n",
       "      <td>DRR240810</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJDB10404</td>\n",
       "      <td>SAMD00239194</td>\n",
       "      <td>Suberoyl bis-hydroxamic acid induced KSHV reac...</td>\n",
       "      <td>Suberoyl bis-hydroxamic acid (SBHA), a histone...</td>\n",
       "      <td>DRS164859</td>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>NIID</td>\n",
       "      <td>DRX230669</td>\n",
       "      <td>...</td>\n",
       "      <td>1042</td>\n",
       "      <td>655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15801</th>\n",
       "      <td>SRR22253895</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA900202</td>\n",
       "      <td>SAMN31682143</td>\n",
       "      <td>Functional and molecular dissection of viral l...</td>\n",
       "      <td>Small, compact genomes confer a selective adva...</td>\n",
       "      <td>SRS15726532</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NARRY KIM LAB, CENTER FOR RNA RESEARCH, INSTIT...</td>\n",
       "      <td>SRX18230429</td>\n",
       "      <td>...</td>\n",
       "      <td>1286</td>\n",
       "      <td>1263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19713</th>\n",
       "      <td>SRR13781372</td>\n",
       "      <td>1</td>\n",
       "      <td>PRJNA704727</td>\n",
       "      <td>SAMN18053879</td>\n",
       "      <td>Respiratory syncytial virus induces alternativ...</td>\n",
       "      <td>we examine the effect of RSV infection in huma...</td>\n",
       "      <td>SRS8317381</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX10166868</td>\n",
       "      <td>...</td>\n",
       "      <td>1318</td>\n",
       "      <td>39991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Long</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ID  kmer_coverage   bioproject     biosample  \\\n",
       "350    SRR14212343              1  PRJNA721322  SAMN18714204   \n",
       "1303   SRR22729431              1  PRJNA912090  SAMN32218795   \n",
       "1484   SRR13781371              1  PRJNA704727  SAMN18053880   \n",
       "3086    ERR5670887              1   PRJEB42052  SAMEA8515329   \n",
       "3730   SRR13781369              1  PRJNA704727  SAMN18053882   \n",
       "7840     DRR240810              1   PRJDB10404  SAMD00239194   \n",
       "15801  SRR22253895              1  PRJNA900202  SAMN31682143   \n",
       "19713  SRR13781372              1  PRJNA704727  SAMN18053879   \n",
       "\n",
       "                                        bioproject_title  \\\n",
       "350    RNA-Seq analysis (long-reads) of human parenta...   \n",
       "1303   Exploring collateral RNA cleavage by LbuCas13a...   \n",
       "1484   Respiratory syncytial virus induces alternativ...   \n",
       "3086   Targeting the m6A RNA modification pathway blo...   \n",
       "3730   Respiratory syncytial virus induces alternativ...   \n",
       "7840   Suberoyl bis-hydroxamic acid induced KSHV reac...   \n",
       "15801  Functional and molecular dissection of viral l...   \n",
       "19713  Respiratory syncytial virus induces alternativ...   \n",
       "\n",
       "                                  bioproject_description   sample_acc  \\\n",
       "350    M238P BRAFV600E mutant melanoma cells sensitiv...   SRS8683487   \n",
       "1303   LbuCas13a exhibits widespread collateral RNA c...  SRS16133703   \n",
       "1484   we examine the effect of RSV infection in huma...   SRS8317380   \n",
       "3086   N6-methyladenosine (m6A) is an abundant intern...   ERS6200127   \n",
       "3730   we examine the effect of RSV infection in huma...   SRS8317378   \n",
       "7840   Suberoyl bis-hydroxamic acid (SBHA), a histone...    DRS164859   \n",
       "15801  Small, compact genomes confer a selective adva...  SRS15726532   \n",
       "19713  we examine the effect of RSV infection in huma...   SRS8317381   \n",
       "\n",
       "      assay_type                                        center_name  \\\n",
       "350    ncRNA-Seq                                                GEO   \n",
       "1303     FL-cDNA                                               LUMC   \n",
       "1484     RNA-Seq                                                GEO   \n",
       "3086       OTHER             NEW YORK UNIVERSITY SCHOOL OF MEDICINE   \n",
       "3730     RNA-Seq                                                GEO   \n",
       "7840     FL-cDNA                                               NIID   \n",
       "15801    RNA-Seq  NARRY KIM LAB, CENTER FOR RNA RESEARCH, INSTIT...   \n",
       "19713    RNA-Seq                                                GEO   \n",
       "\n",
       "        experiment  ... avgspotlen mbases  biosamplemodel_sam  \\\n",
       "350    SRX10578990  ...        698  11980                 NaN   \n",
       "1303   SRX18690947  ...        382    537               Human   \n",
       "1484   SRX10166867  ...       1508  33362                 NaN   \n",
       "3086    ERX5386380  ...       1669    753                 NaN   \n",
       "3730   SRX10166865  ...       1481  27528                 NaN   \n",
       "7840     DRX230669  ...       1042    655                 NaN   \n",
       "15801  SRX18230429  ...       1286   1263                 NaN   \n",
       "19713  SRX10166868  ...       1318  39991                 NaN   \n",
       "\n",
       "       collection_date_sam location_count location_avg_coverage  \\\n",
       "350                    NaN              0                   NaN   \n",
       "1303                   NaN              0                   NaN   \n",
       "1484                   NaN              0                   NaN   \n",
       "3086                   NaN             43                   1.0   \n",
       "3730                   NaN              0                   NaN   \n",
       "7840                   NaN              0                   NaN   \n",
       "15801                  NaN              0                   NaN   \n",
       "19713                  NaN              0                   NaN   \n",
       "\n",
       "      location_max_coverage location_min_coverage location_med_coverage  \\\n",
       "350                     NaN                   NaN                   NaN   \n",
       "1303                    NaN                   NaN                   NaN   \n",
       "1484                    NaN                   NaN                   NaN   \n",
       "3086                    1.0                   1.0                   1.0   \n",
       "3730                    NaN                   NaN                   NaN   \n",
       "7840                    NaN                   NaN                   NaN   \n",
       "15801                   NaN                   NaN                   NaN   \n",
       "19713                   NaN                   NaN                   NaN   \n",
       "\n",
       "      reads_type  \n",
       "350         Long  \n",
       "1303        Long  \n",
       "1484        Long  \n",
       "3086        Long  \n",
       "3730        Long  \n",
       "7840        Long  \n",
       "15801       Long  \n",
       "19713       Long  \n",
       "\n",
       "[8 rows x 51 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_SJ_df = tags_and_dataframes_dict[0][1]\n",
    "human_altsjdf = alt_SJ_df.query(\"organism == 'Homo sapiens'\")\n",
    "print(f\"There's {len(human_altsjdf)} human datasets in the 'alt_SJ' results.\")\n",
    "human_long_reads_for_alt_SJ_df = human_altsjdf.query(\"reads_type == 'Long'\")\n",
    "print(f\"Out of those {len(human_altsjdf)}, there's {len(human_long_reads_for_alt_SJ_df)} 'long-read' ones.\")\n",
    "human_long_reads_for_alt_SJ_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be15e99c-79b1-416a-b58a-d46b6776c026",
   "metadata": {},
   "source": [
    "We see eight datasets have 'Long' for the reads type."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5435eba7-3480-4ad1-8a1e-0acd003a0c66",
   "metadata": {},
   "source": [
    "The produced dataframe could be used to consider possibe datasets to investigate further.\n",
    "\n",
    "-----\n",
    "\n",
    "Enjoy!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
